flexget/utils/serialization.py

```diff
@@ -0,0 +1,128 @@
+    if isinstance(value, datetime.datetime):
+        return DateTimeSerializer.serialize(value)
+    if isinstance(value, datetime.date):
+        return DateSerializer.serialize(value)
+    if isinstance(value, (str, int, float, type(None))):
+        return value
+    raise TypeError(f'`{value!r}` of type {type(value)!r} is not serializable')
+
+
+def deserialize(value):
+    if isinstance(value, dict):
+        if all(key in value for key in ('serializer', 'version', 'value')):
+            return _registry()[value['serializer']].deserialize(value)
+        return {k: deserialize(v) for k, v in value.items()}
+    if isinstance(value, list):
+        return [deserialize(v) for v in value]
+    return value
+
+
+class Serializable(ABC):
+    @abstractmethod
+    def _serialize(self):
+        """This method should be implemented to return a plain python datatype which is json serializable."""
+        pass
+
+    def serialize(self):
+        """Returns a json serializable form of this class, with all information needed to deserialize."""
+        return {
+            'serializer': self.serializer_name(),
+            'version': self.serializer_version(),
+            'value': self._serialize(),
+        }
+
+    @classmethod
+    @abstractmethod
+    def _deserialize(cls, data, version):
+        """Returns an instance of this class, recreated from the serialized form."""
+        pass
+
+    @classmethod
+    def deserialize(cls, data):
+        return cls._deserialize(data['value'], data['version'])
+
+    @classmethod
+    def serializer_name(cls):
+        return cls.__name__
+
+    @classmethod
+    def serializer_version(cls):
+        return 1
+
+    def dumps(self):
+        return json.dumps(self.serialize())
+
+    @classmethod
+    def loads(cls, data):
+        return cls.deserialize(json.loads(data))
+
+
+class DateSerializer(Serializable):
+    @classmethod
+    def serializer_name(cls):
+        return 'date'
+
+    def _serialize(self):
+        pass
+
+    @classmethod
+    def serialize(cls, value):
+        return {
+            'serializer': cls.serializer_name(),
+            'version': cls.serializer_version(),
+            'value': value.strftime(DATE_FMT),
+        }
+
+    @classmethod
+    def _deserialize(cls, data, version):
+        return datetime.datetime.strptime(data, DATE_FMT).date()
+
+
+class DateTimeSerializer(Serializable):
+    @classmethod
+    def serializer_name(cls):
+        return 'datetime'
+
+    def _serialize(self):
+        pass
+
+    @classmethod
+    def serialize(cls, value):
+        return {
+            'serializer': cls.serializer_name(),
+            'version': cls.serializer_version(),
+            'value': value.strftime(ISO8601_FMT),
+        }
+
+    @classmethod
+    def _deserialize(cls, data, version):
+        return datetime.datetime.strptime(data, ISO8601_FMT)
+
+
+_registry_cache = None
+
+
+def _registry():
+    global _registry_cache
```

Comment:
If you initialize `_registry_cache` as dict straight away you don't need to use global or make checks and initialize it as adding to dict works fine without global.

Reply:
True. I guess I was thinking about telling the difference from when there were no subclasses defined, but that situation doesn't matter at all.

Reply:
Hmm. I know this works fine, but it still irks me that PyCharm highlights this. I'm tempted to leave the global in there just to avoid it. 😛 
![image](https://user-images.githubusercontent.com/187133/72620757-ea827e00-390d-11ea-8cd0-7e3dff983711.png)

Reply:
Hmm. Actually, it does not work fine. Is this a new behavior in later pythons?

Reply:
Doh. Just me being silly. I was assigning to that variable rather than mutating it.

Reply:
> If you initialize `_registry_cache` as dict straight away you don't need to use global or make checks and initialize it as adding to dict works fine without global.

Still needs to make checks and initialize, otherwise it gets initialized at the module level, which is before all the other subclasses of Serializable have been created.